### ISSUE TITLE
model.preidct() returns nan values ? 

### ISSUE THREAD

**Issue by ArunVishw** [Neerajj9/CSRNet-keras#22]
I used this code and after training when I run inference.ipynb 
model.predict() is giving NAN values
![Screenshot (91)](https://user-images.githubusercontent.com/45504815/75914608-0a004680-5e7b-11ea-973a-cdb1a471e484.png)

**Comment by azelk** [comment]
I have the same problem. At some point in training the loss always becomes nan and so does the predictions result.

I've tried removing from dataset images where the count is 0, normalized data in different ways, tried changing sgd to adam, and nothing helps.

Did you or anyone solve this?

**Comment by azelk** [comment]
The code also uses Euclidean distance loss and its derivative becomes infinite at 0

**Comment by ArunVishw** [comment]
> Hi, I had the same problem when at some point the training loss becomes nan and so does the predictions result.
> 
> I've tried removing from dataset the cases where the count is 0, normalized data in different ways, tried changing sgd to adam and it didn't help much, but then I've increased the learning rate to 0.001 and it helped! So try this too
> 
> Perhaps the gradient was already so small that with learning rate ~1e-9 it simply became 0.

Training loss is not 0 in y case...
What i found out is that all parameters calculated are NaN.
I opened the model_A_weights.h5 file and all bias and kernel parameters are NaN.

**Comment by DanyaAlfageh** [comment]
Hello I am also facing this issue has anyone found a solution yet ?

**Comment by DanyaAlfageh** [comment]
@azelk I am fairly new at this but I changed the optimizer to Adam and it produced values is that okay ?

**Comment by azelk** [comment]
@DanyaAlfageh yes, quite often it does help but in my case it has only partially fixed the problem (i.e. it still produced NaN values from time to time, but I have a different dataset / counting problem which might have contributed)

**Comment by ArunVishw** [comment]
> @azelk I am fairly new at this but I changed the optimizer to Adam and it produced values is that okay ?

Can you tell me how have you calculated distance loss ?
whether you are using the same euclidean_distance_loss as it is ?

**Comment by DanyaAlfageh** [comment]
@ArunVishw I am using the below. I am fairly new to this so I am not 100% sure of the correctness of what I did
```
            adam= adam(lr = 1e-6, decay = (5*1e-4))
            model.compile(loss='binary_crossentropy',optimizer=adam,metrics=['mse'])
```

**Comment by DanyaAlfageh** [comment]
for the loss function it is defined as follows:
```
def euclidean_distance_loss(y_true, y_pred):
    # Euclidean distance as a measure of loss (Loss function)    return K.sqrt(K.sum(K.square(y_pred - y_true), axis=-1))
```
My question is how will it get y_true and y_pred when they are never passed to the fucntion ? is it why it is producing Nan ?

**Comment by ArunVishw** [comment]
> for the loss function it is defined as follows:
> 
> ```
> def euclidean_distance_loss(y_true, y_pred):
>     # Euclidean distance as a measure of loss (Loss function)    return K.sqrt(K.sum(K.square(y_pred - y_true), axis=-1))
> ```
> 
> My question is how will it get y_true and y_pred when they are never passed to the fucntion ? is it why it is producing Nan ?

No the reason NaN values were produced is due to use of euclidean_loss_function(), 
if we use use mean_sqaured_error or binary_crossentropy as loss function it will work just fine.
Using adam optimizer is recommended..

And (y_true, y_pred) gets passed automatically to loss function, check out keras documentation of loss function for more details.
